back/app/services/user_reduce_service.rb

```diff
@@ -0,0 +1,142 @@
+    skip_users ||= User.admin
+    scope = User.where.not id: skip_users
+
+    project_sets = compute_project_sets scope
+    loop do
+      can_pick_empty_set = true
+      picked_project_sets = Set.new
+      picked_users = []
+      while (project_set = pick_next_project_set! project_sets, exclude: picked_project_sets, can_pick_empty_set: can_pick_empty_set)
+        picked_project_sets ||= project_set
+        picked_project_sets |= project_set
+        picked_users += [project_sets[project_set].pop]
+        can_pick_empty_set = false if project_set.empty?
+      end
+
+      return if picked_users.size < 2
+
+      merge! picked_users, project_sets: project_sets
+
+      return if timeout && (Time.now - t_start) > timeout
+    end
+  end
+
+  def merge!(users, project_sets: nil)
+    # Merges all other users into the first user. Deletes the other users.
+    merged_user = pick_and_pop_user_for_merge! users, project_sets: project_sets
+    users_to_merge = users
+
+    occurences = {}
+    uuid_columns.select do |table_name, column_names|
+      column_names.each do |column_name|
+        ActiveRecord::Base.connection.execute(
+          <<-SQL.squish
+            SELECT #{column_name}
+            FROM #{table_name}
+            WHERE #{column_name} IN (#{users_to_merge.map { |u| "'#{u.id}'" }.join(',')})
+          SQL
+        ).present?
+      end
+      occurences[table_name] = column_names if column_names.present?
+    end
+
+    ActiveRecord::Base.transaction do
+      occurences.each do |table_name, columns|
+        columns.each do |column|
+          query = <<-SQL.squish
+            UPDATE #{table_name}
+            SET #{column} = \'#{merged_user.id}\'
+            WHERE #{column} IN (#{users_to_merge.map { |u| "'#{u.id}'" }.join(',')})
+          SQL
+          ActiveRecord::Base.connection.execute query
+        end
+      end
+      users_to_merge.each(&:destroy!)
+    end
+  end
+
+  private
+
+  def compute_project_sets(scope)
+    # Project sets are hashes where the keys are
+    # sets of project IDs and the values are
+    # users who participated in exacly just those
+    # projects.
+
+    picked_project_sets = {}
+    participants_service = ParticipantsService.new
+
+    context_participants = Project.all.map do |project|
+      participants = participants_service.project_participants project
+      [project.id, participants.map(&:id)]
+    end.to_h
+    context_participants['global'] = participants_service.initiatives_participants(Initiative.all).map(&:id)
+
+    scope.each do |user|
+      set = Set.new
+      context_participants.each do |context, participant_ids|
+        set.add context if participant_ids.include? user.id
+      end
+      picked_project_sets[set] ||= []
+      picked_project_sets[set] += [user]
+    end
+    picked_project_sets.values.each(&:shuffle!)
+    picked_project_sets
+  end
+
+  def pick_next_project_set!(project_sets, exclude: Set.new, can_pick_empty_set: true)
```

Comment:
I've not seen this use of `exclude` in method parameters before, and I can't find any documentation on it. Could you point me at some documentation?